### PR TITLE
Web UI fixes

### LINKED ISSUE #2739: No Confirmation if User Navigates Away From Wire Graph
If user navigates to some other service without saving a wire graph, kura does not ask user if he wants to save a graph or not - and graph gets returned to previous state. 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Wire Graph, create several componetns and without applying the changes, switch view to some other service (e.g. example publisher).
3. Observ the lack of info message from kura regarding the newly created wire graph
4. Return to the wire graph - there are no components

**Expected behavior**
If you want to switch to another view, Kura should always ask you if you want to save/apply the new changes. 

**Screenshot:**
/

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Additional info: /

### LINKED ISSUE #2746: WireGraph Can be Saved Although CamelConsumer/Producer Fields Not Filled
Although user does not fill the mandatory parameters into the CamelConsumer/CamelProducer, Kura does not throw an error (mandatory fields are "ID" and "EndpointURI"). 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to WireGraph and select CamelConsumer/CamelProducer
3. Without populating the mandatory fields, try to save the graph
4. Kura does not report an error when trying to do so.

**Expected behavior**
If mandatory fields in CamelConsumer/Producer are not filled, graph should not be saved.

**Screenshots**
/

**Target Environment (please complete the following information):**
 - Board [e.g. Raspberry Pi 3]
 - OS version: [e.g. provide the result of `uname -a`]
 - Additional info: 

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Additional info: /

### LINKED ISSUE #2748: "Phantom" Component When Ignoring Error in Wire Graph
If user tries to create a component with non-unique name and first presses "ESC" key on keyboard before closing the error message, the "phantom" component stays in the graph, you can even save the graph if you add another real component. 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Wire Graph
3. Drag and drop a component (e.g. publisher)
4. Enter a non-unique name so error appears
5. Press "ESC" key on keyboard
6. close error message
7. Observe the "phantom" component

**Expected behavior**
There should be no "phantom" component regardless of what user does. 

**Screenshots**
Screenshot 2020-04-01 at 12 24 48

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Additional info: /

### LINKED ISSUE #2753: No File Check When Uploading Channels to Asset in Wire Graph
If user tries to upload a non-CSV Channels file in Wire Graph, Kura reports an error, which is OK, but the output could be messy (it depends on the type of the file and length of it - in my case I uploaded a .txt file with a lot of contents) - see the screenshot bellow. 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Packages, install Wires-development-package from Eclipse Marketplace
3. Go to Wire Graph, create a wireAsset comonent and add a channel to it
4. Try to upload a random .txt file via "Upload channels" button
5. Observe the error message

**Expected behavior**
I suggest that Kura checks if the uploaded file has a .CSV extension (although user can have an inappropriate .cvs file), this way we would take care of a lot errors. 

**Screenshot:**
Screenshot 2020-04-01 at 14 38 00

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Additional info: /

### LINKED ISSUE #2757: True/False Icons in Asset Must be Double Clicked
If user creates a wire asset in Wire graph, the "enabled" and "listen" checkboxes have to be doubl eclicked to be selected/deselected insteadn of being clicked once to achieve the desired status. This is because the fist click actually selects the channel (although it looks like selected when user hoovers above) and the second click actually selects/deselects the option. 
Is this the corect behaviour or should we change this so the checkbox gets changed after the first click? 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Packages, install Wires-development-package from Eclipse Marketplace
3. Go to Wire Graph, create a wireAsset comonent and add a channel to it
4. Try to select/deselect the "Enabled" and "Listen" options with one click - it is not possible

**Expected behavior**
See the suggestion above. 

**Screenshot:**

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Additional info: /

### LINKED ISSUE #2758: "Last Page" Button in Assets Not Working
If user creates more than 15 channels in one asset, the "Last page" button is not active if user is on first page. "First page" button works fine, though. 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Packages, install Wires-development-package from Eclipse Marketplace
3. Go to Drivers and assets, create a driver with asset
4. Add at least 16 channels to this asset
5. Observe the "Last page" button when on first page - it does not work

**Expected behavior**
"Last page" button should be active if there are more pages than one and user is on page one. 

**Screenshot:**
Screenshot 2020-04-02 at 16 13 19

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Additional info: /.

### PR DESCRIPTION
**Pull request by nicolatimeus**
Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

**Related Issues:**

Closes #2757 
Closes #2753
Closes #2748
Closes #2739
Closes #2758 
Closes #2746

### CODE REVIEW

**Comment by LeoNerdoG** [comment]
Hey @nicolatimeus,
I have checked this PR and the results are:
 - 2757: OK
 - 2753: There are two error messages if .csv file is not valid:
    - "Empty CSV file" and
    - "Validation exception on CSV file."
I would expect that if .csv file is actually empty that first message is shown and if .csv file is not valid - the other message. 
Screenshot 2020-06-05 at 13 29 10

 - 2748: OK
 - 2739: OK
 - 2758: There is an additional button left:
Screenshot 2020-06-05 at 14 16 09

 - 2746: OK

**Comment by nicolatimeus** [comment]
Regarding 2758, the behaviour of the buttons is the following:

* The `>` button shows next page.
* The `>>` button is a fast-forward button, before this PR it advanced the view by 1000 rows, for this reason it was always disabled unless a lot of channels are defined. After this PR it advances by 6 pages (90 rows), hopefully making it useful in more cases. 
* The `>|` button moves the view to last page, this button is introduced in this PR.

GWTbootstrap does not support a "fast-backwards" button.

**Comment by LeoNerdoG** [comment]
Hey,
ok, noted. I have an idea though if the "fast backwards" button is not available - should we add a free netry field for the page number? Do you think that this would be something interesting for the end user?

**Comment by MMaiero** [comment]
Do we need the fast forward button? 
I guess it can be disabled. If so, maybe it's simpler for the user when there is a limited number of buttons: a button to go to the next page and one to the last.

**Comment by MMaiero** [comment]
I have also noticed that the move to next page and the move to last page act a little bit weirdly, at least to me.
Supposing to have 19 channels, there should be 2 "pages". I would expect that clicking the next page and the last page button will output the same view. Instead they output something like this:
Screenshot 2020-06-08 at 14 21 31

Screenshot 2020-06-08 at 14 21 39

**Comment by LeoNerdoG** [comment]
Hey @MMaiero,
you are right. It looks like Kura adapts to the situtation - if you have e.g. 3 pages of assets, then first page contains 1-15, second 15-30 and third on 25-40 (if you have 40 channels).
Should this be fixed in this PR? Or should new issue be opened and then sepparate PR to fix it?

**Comment by MMaiero** [comment]
Don't know. Depends on the effort that @nicolatimeus can estimate. Personally, I would fix in this PR, if possible

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/wires/WiresDialogs.ui.xml [hunk 1]
```diff
@@ -23,7 +23,7 @@
     </ui:style>
 
     <b:Container fluid="true">
-        <b:Modal closable="false" dataKeyboard="true" b:id="select-asset-modal" ui:field="selectAssetModal"
+        <b:Modal dataKeyboard="true" b:id="select-asset-modal" ui:field="selectAssetModal"
```
Comment (MMaiero) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/wires/WiresDialogs.ui.xml]: Probably we need to align with the changes made in #2965

**Comment by LeoNerdoG** [comment]
Hey @nicolatimeus,
I have checked this PR after the latest commit and the status not is:
 -  #2757: OK
 -  #2753: still issue. If the file starts with "-" (dash), then the error message is:
Screenshot 2020-07-21 at 10 36 06

 -  #2748: OK
 -  #2739: OK
 -  #2758: OK
 -  #2746: OK

**Review by MMaiero — APPROVED** [review]

**Comment by nicolatimeus** [comment]
Hi @LeoNerdoG , this PR does not change the csv validation logic or the error reporting ui. 
It only includes a best effort change that prevents the user from selecting files that do not have the csv extension, as suggested in #2753.
Maybe we can improve the CSV validation and error reporting as part of other PRs, I think that will require some more work.

**Comment by LeoNerdoG** [comment]
Hey @nicolatimeus,
then I suggest we merge this PR but we keep open the issue #2753 until we figure out a better solution, what do you think?

**Pull request merged**